### PR TITLE
fix: hold Expense option is displayed on expenses on a closed report

### PR DESCRIPTION
**Pull request by nkdengineer**
### Fixed Issues

$ https://github.com/Expensify/App/issues/42202
PROPOSAL: https://github.com/Expensify/App/issues/42202#issuecomment-2114041222

### Tests

1. Set a workspace without an approval flow
2. Invite a employee
3. As the employee - Submit 2 expenses to the workspace
4. Submit the report - Since there's no approval workflow - the report should auto-close
5. Navigate to a expense of the closed report
6. Click on the 3-dot menu
7. Verify that: Hold menu option will not be present.

- [x] Verify that no errors appear in the JS console

### QA Steps

1. Set a workspace without an approval flow
2. Invite a employee
3. As the employee - Submit 2 expenses to the workspace
4. Submit the report - Since there's no approval workflow - the report should auto-close
5. Navigate to a expense of the closed report
6. Click on the 3-dot menu
7. Verify that: Hold menu option will not be present.

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

android

</details>

<details>
<summary>Android: mWeb Chrome</summary>

android-mweb

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/161821005/61e1b76c-9be2-4436-8eec-ef0e5da75fb9

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/Expensify/App/assets/161821005/c1525872-b5ec-424e-b651-764d20657fde

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

web

</details>

<details>
<summary>MacOS: Desktop</summary>

desktop

</details>

### CODE REVIEW

File: src/libs/actions/IOU.ts [hunk 1]
```diff
@@ -6679,6 +6679,7 @@ function putOnHold(transactionID: string, comment: string, reportID: string) {
                 comment: {
                     hold: null,
                 },
+                errors: ErrorUtils.getMicroSecondOnyxError('report.genericAddCommentFailureMessage'),
```
Comment (rojiphil) [inline src/libs/actions/IOU.ts:6682]: The mentioned error genericAddCommentFailureMessage -> `Unexpected error while posting the comment, please try again later.` does not sound correct here. Can you please use a generic hold related error message along with Spanish translation?

File: src/libs/actions/IOU.ts [hunk 2]
```diff
@@ -6740,6 +6741,7 @@ function unholdRequest(transactionID: string, reportID: string) {
             key: `${ONYXKEYS.COLLECTION.TRANSACTION}${transactionID}`,
             value: {
                 pendingAction: null,
+                errors: ErrorUtils.getMicroSecondOnyxError('report.genericAddCommentFailureMessage'),
```
Comment (rojiphil) [inline src/libs/actions/IOU.ts:6744]: Same here too. Let us use a generic hold error message here.

File: src/libs/actions/IOU.ts (hunk 1 above)
Comment (rojiphil) [reply src/libs/actions/IOU.ts:6682]: Preferably in iou->error makes more sense to me.

**Comment by nkdengineer** [comment]
@rojiphil i added new generic error, and ask for Spanish in [here](https://expensify.slack.com/archives/C01GTK53T8Q/p1716887419042779)
please check again

**Comment by rojiphil** [comment]
> @rojiphil i added new generic error, and ask for Spanish in [here](https://expensify.slack.com/archives/C01GTK53T8Q/p1716887419042779) please check again

@nkdengineer First we need to get the English text reviewed with marketing team. Only then we have to post in the slack channel. We can do this here in the PR itself.

**Comment by rojiphil** [comment]
@Expensify/marketing We want to display an error message when BE sends an error response to the `Hold Expense` API and `Unhold Expense` API.

Does the following error messages in English look good?
`Unexpected error while holding the expense. Please try again later.`
`Unexpected error while unholding the expense. Please try again later.`

**Review by pecanoro — CHANGES_REQUESTED** [review]

**Comment by jamesdeanexpensify** [comment]
@rojiphil @pecanoro I need to get the English copy approved first. Will bring this out into Slack and then circle back here. Thanks!

**Comment by jamesdeanexpensify** [comment]
Actually one question - do we know any reason for these errors, so we can tell the user what's going on and be more specific about how to fix the issue?

**Comment by rojiphil** [comment]
>  do we know any reason for these errors, so we can tell the user what's going on and be more specific about how to fix the issue?

@jamesdeanexpensify  Well! No. We just wanted to capture any unexpected error as currently there is no error message displayed if the API request failed.
And oh! I just noticed a generic error message i.e. `Unexpected error, please try again later.`  in the translations [here](https://github.com/Expensify/App/blob/2c8b1d50f4ddb0cf8cc60ee786efb2fc57723a69/src/languages/en.ts#L726). This has been used for unexpected errors during IOU requests like `CancelPayment`, `SubmitReport`
I think this may be good enough. 
What do you think?

**Comment by jamesdeanexpensify** [comment]
Getting some quick internal thoughts! Back in a bit.

**Comment by jamesdeanexpensify** [comment]
Let's go with:

> Unexpected error while holding the expense. Please try again later.
> Unexpected error while taking the expense off hold. Please try again later.

**Comment by nkdengineer** [comment]
@rojiphil @pecanoro @jamesdeanexpensify i updated, please check again 🙏

**Comment by jamesdeanexpensify** [comment]
I **think** the next step is just @pecanoro taking a look at the Spanish translations of:

> Unexpected error while holding the expense. Please try again later.
> Unexpected error while taking the expense off hold. Please try again later.

**Comment by rojiphil** [comment]
> Waiting for @rojiphil to leave a review so I can leave my final one!

Oh! Sorry. Completely missed this.

**Comment by rojiphil** [comment]
@pecanoro Here are the computer generated translations for the approved wordings. 
Does this sound correct in Spanish?

English: Unexpected error while holding the expense. Please try again later.
Spanish: Error inesperado al realizar el gasto. Vuelva a intentarlo más tarde.

English: Unexpected error while taking the expense off hold. Please try again later.
Spanish: Error inesperado al retirar el gasto de la espera. Vuelva a intentarlo más tarde.

**Comment by rojiphil** [comment]
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] Android: Native
    - [ ] Android: mWeb Chrome
    - [ ] iOS: Native
    - [ ] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [ ] MacOS: Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [ ] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [ ] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/3004659/79392abd-25cc-44b0-95ff-e4abcd2d558c

</details>

<details>
<summary>Android: Native</summary>



</details>

<details>
<summary>Android: mWeb Chrome</summary>



</details>

<details>
<summary>iOS: Native</summary>



</details>

<details>
<summary>iOS: mWeb Safari</summary>



</details>

<details>
<summary>MacOS: Desktop</summary>



</details>

**Review by rojiphil — COMMENTED** [review]
Code LGTM. Tests well on web platform. Looking forward to complete the checklist once the following comments are addressed.

@nkdengineer Here are few comments for addressal.
1. We can remove Step 7 in the `Tests` section as `Hold` menu option will not be present. 
2. Please add test cases and screenshots/videos to demonstrate display of error messages due to failure in `Hold` and `Unhold` API requests.
~3. While we wait for the correct translations, let us, meanwhile, use the computer generated ones as mentioned [here](https://github.com/Expensify/App/pull/42574#issuecomment-2137804728)~
3rd point can be ignored as it is going to be updated anyway in the final review as mentioned [here](https://github.com/Expensify/App/pull/42574#issuecomment-2137758484)

**Comment by nkdengineer** [comment]
@rojiphil For the error case that I mentioned in my proposal, we have a BE issue here. When device 1 hold a request in offline and device 2 hold the request online, after device 1 go online `HoldRequest` API doesn't return jsonCode with error although the OnyxData is empty and when we reset cache, the error hold action disappear.

I think we can create a separate internal issue for this case that will return as an error API if the request is already held.

cc @dangrous 
 
https://github.com/Expensify/App/assets/161821005/d5306d51-1cbd-47d6-8674-7bc63e124b6c

**Comment by pecanoro** [comment]
I already left suggestions for the translations so we are covered in that sense

**Comment by pecanoro** [comment]
Oh wait forget it, I though I did, but it seems I didn't

File: src/languages/es.ts [hunk 3]
```diff
@@ -718,6 +718,8 @@ export default {
             invalidSplit: 'La suma de las partes debe ser igual al importe total.',
             invalidSplitParticipants: 'Introduce un importe superior a cero para al menos dos participantes.',
             other: 'Error inesperado, por favor inténtalo más tarde.',
+            genericHoldExpenseFailureMessage: 'Error inesperado al bloquear el gasto, por favor inténtela de nuevo más tarde.',
+            genericUnholdExpenseFailureMessage: 'Error inesperado al desbloquear el gasto, por favor inténtela de nuevo más tarde.',
```
Comment (pecanoro) [inline src/languages/es.ts:722]:
```suggestion
            genericHoldExpenseFailureMessage: 'Error inesperado al bloquear el gasto, por favor inténtalo de nuevo más tarde.',
            genericUnholdExpenseFailureMessage: 'Error inesperado al desbloquear el gasto, por favor inténtalo de nuevo más tarde.',
```

**Review by pecanoro — CHANGES_REQUESTED** [review]

**Comment by rojiphil** [comment]
> For the error case that I mentioned in my proposal, we have a BE issue here.

Thanks for pointing this out. I did another test which involved holding the expense offline in first device while submitting and closing the report from second device. In this case too, the `HoldRequest` API should have failed when the first device came online. But it didn’t. Here is a test video demonstrating this. 

So, we have two options to take this forward:
1. Make this issue `internal`, fix in BE and, then, test and close this issue. Or
2. As suggested before, complete the checklist here and create separate internal issue for BE. 

I also think 2. may be better.

@dangrous Please advice 

https://github.com/Expensify/App/assets/3004659/4fd92e42-101c-4812-985d-b145b0c33977

**Comment by pecanoro** [comment]
@dangrous I also think we should do option 2, what do you think?

**Review by rojiphil — APPROVED** [review]
While we await the final decision from @dangrous, I have verified that the original issue is resolved on all platforms and have completed the checklist.
However, if we go with option 1 I will rerun the checklist.

**Review by pecanoro — APPROVED** [review]

**Comment by dangrous** [comment]
Yep, I agree - let's make a separate issue for that BE issue.

Will review/merge this in one sec!

**Review by dangrous — APPROVED** [review]

**Comment by dangrous** [comment]
Did we make the internal issue for https://github.com/Expensify/App/pull/42574#issuecomment-2142499095 yet? I did not, but I can shortly if not!

**Comment by nkdengineer** [comment]
> Did we make the internal issue for https://github.com/Expensify/App/pull/42574#issuecomment-2142499095 yet? I did not, but I can shortly if not!

We didn't. I think you can create an internal issue for it.

**Comment by dangrous** [comment]
Internal issue created! https://github.com/Expensify/Expensify/issues/406394 . Screenshot below for non-internal folks.

image

File: src/libs/actions/IOU.ts [hunk 4]
```diff
@@ -6749,6 +6749,7 @@ function putOnHold(transactionID: string, comment: string, reportID: string) {
                 comment: {
                     hold: null,
                 },
+                errors: ErrorUtils.getMicroSecondOnyxError('iou.genericHoldExpenseFailureMessage'),
```
[the remark this answers is not in the document]
Comment (dangrous) [reply src/libs/actions/IOU.ts:6752]: Ah, thank you!

**Pull request merged**